common/util/vector_tree.h

```diff
@@ -198,6 +200,172 @@ class _VectorTreeImpl {
+  using Base::crend;
+  using Base::rbegin;
+  using Base::rend;
+
+  using Base::back;
+  using Base::front;
+  using Base::operator[];
+  using Base::at;
+
+  using Base::empty;
+  using Base::max_size;
+  using Base::size;
+
+  using Base::emplace_back;
+  using Base::push_back;
+
+  using Base::emplace_front;
+  using Base::push_front;
+
+  using Base::emplace;
+  using Base::insert;
+
+  using Base::clear;
+  using Base::erase;
+  using Base::pop_back;
+  using Base::pop_front;
+
+  using Base::assign;
+  using Base::operator=;
+  using Base::swap;
+
+  using Base::capacity;
+  using Base::reserve;
+  using Base::resize;
+
+ protected:
+  // ContainerProxy interface
+
+  container_type& underlying_container() { return container_; }
+  const container_type& underlying_container() const { return container_; }
+
+  void ElementsInserted(iterator first, iterator last) {
+    for (auto& child : iterator_range(first, last)) {
+      child.parent_ = &node_;
+    }
+  }
+
+  // Unused:
+  // void ElementsBeingRemoved(iterator first, iterator last)
+
+  // Unused:
+  // void ElementsBeingReplaced()
+
+  void ElementsWereReplaced() {
+    for (auto& child : container_) {
+      child.parent_ = &node_;
+    }
+  }
+
+ private:
+  // Hide constructors and assignments from the world. This object is created
+  // and assigned-to only in VectorTree.
+
+  explicit VectorTreeChildrenList(VectorTreeType& node) : node_(node) {}
+
+  // Construction requires parent node reference.
+  VectorTreeChildrenList(const VectorTreeChildrenList&) = delete;
+
+  VectorTreeChildrenList(VectorTreeType& node,
+                         const VectorTreeChildrenList& other)
+      : node_(node), container_(other.container_) {
+    for (auto& child : container_) {
+      child.parent_ = &node;
+    }
+  }
+
+  VectorTreeChildrenList& operator=(const VectorTreeChildrenList& other) {
+    container_ = other.container_;
+    for (auto& child : container_) {
+      child.parent_ = &node_;
+    }
+    return *this;
+  }
+
+  // Construction requires parent node reference.
+  VectorTreeChildrenList(VectorTreeChildrenList&&) = delete;
+
+  VectorTreeChildrenList(VectorTreeType& node,
+                         VectorTreeChildrenList&& other) noexcept
+      : node_(node), container_(std::move(other.container_)) {
+    // Note: `other` is not notified about the change because it ends up in
+    // undefined state as a result of the move.
+    for (auto& child : container_) {
+      child.parent_ = &node;
+    }
+  }
+
+  VectorTreeChildrenList& operator=(VectorTreeChildrenList&& other) noexcept {
+    // Note: `other` is not notified about the change because it ends up in
+    // undefined state as a result of the move.
+    container_ = std::move(other.container_);
+    for (auto& child : container_) {
+      child.parent_ = &node_;
+    }
+    return *this;
+  }
+
+  // Reference to a VectorTree node in which this object represents a list of
+  // children.
+  VectorTreeType& node_;
```

Comment:
One minor concern is that relative to the previous implementation, this member increases the size of a node by an extra pointer.

Reply:
Yes, unfortunately. Alternatives I've tried:
- `VectorTreeChildrenList` as a reference-like class returned from `Children()`. Problems: requires two classes (for const and non-const references) and use of `auto` without `&` everywhere.
- `offsetof` used to calculate pointer to node from pointer to `children_` (i.e. from `this` inside `VectorTreeChildrenList`). Problem: `offsetof` requires stanadard layout class, while both value type and container can have non-standard layout (e.g. with `GLIBCXX_DEBUG` std::vector is not standard-layout).

Reply:
Maybe for now, we should add a TODO here for later reducing the footprint.